{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0d023709-2e3e-40a4-b4bc-770c215c7603",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Introduction\n",
    "* **작성자**: 박소희\n",
    "* **본 주피터 코드 목적**: GEO167186 RNA-seq의 DEG 분석을 위함\n",
    "* **데이터셋**: GSE167186\n",
    "* **데이터셋 설명**:\n",
    "    * Experiment type: Expression profiling by high throughput sequencing\n",
    "    * Platform: Illumina HiSeq 4000\n",
    "    * Experiment molecule: total RNA\n",
    "* **샘플 설명**:\n",
    "    * 평균 20세 (Young): 19명\n",
    "    * 평균 75세 (Old) non-sarcopenic: 29명\n",
    "    * 평균 75세 (Old) sarcopenic: 24명\n",
    "* **참고 논문**: https://www.aging-us.com/article/204435/text\n",
    "* **데이터셋 링크**: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE167186\n",
    "* **PyDESeq2 참고 코드**: https://pydeseq2.readthedocs.io/en/latest/auto_examples/plot_minimal_pydeseq2_pipeline.html\n",
    "* **Enrichr**: https://maayanlab.cloud/Enrichr/enrich#"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4188429b-0fd7-41e7-8f0f-7325efcb9ee0",
   "metadata": {},
   "source": [
    "# Data collection\n",
    "* DEG 분석을 위해 **Count table**과 샘플의 그룹 정보가 포함된 **Metadata**가 필요합니다.\n",
    "* 아래에서 **Count table**과 **Metadata table**을 불러와 형태를 확인합니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "12aaf094-945d-4683-abf7-661ba27eae91",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/soheepark/anaconda3/envs/pydeseq2/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import pickle\n",
    "import os\n",
    "os.chdir('/Users/soheepark/03-GEO근감소/Data/gse167186/')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2bac407-cc67-49f8-a28e-83c16af8fb0a",
   "metadata": {},
   "source": [
    "먼저 **Count table을 확인**해봅니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bd8b12ea-057f-4495-93dc-594c1d4f8873",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(36694, 76)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X_10</th>\n",
       "      <th>X_11</th>\n",
       "      <th>X_13</th>\n",
       "      <th>X_14</th>\n",
       "      <th>X_15</th>\n",
       "      <th>X_16</th>\n",
       "      <th>X_17</th>\n",
       "      <th>X_18</th>\n",
       "      <th>X_19</th>\n",
       "      <th>X_1</th>\n",
       "      <th>...</th>\n",
       "      <th>X_75</th>\n",
       "      <th>X_76</th>\n",
       "      <th>X_77</th>\n",
       "      <th>X_78</th>\n",
       "      <th>X_79</th>\n",
       "      <th>X_7</th>\n",
       "      <th>X_80</th>\n",
       "      <th>X_81</th>\n",
       "      <th>X_8</th>\n",
       "      <th>X_9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>20</td>\n",
       "      <td>35</td>\n",
       "      <td>14</td>\n",
       "      <td>22</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>22</td>\n",
       "      <td>35</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>18</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          X_10  X_11  X_13  X_14  X_15  X_16  X_17  X_18  X_19  X_1  ...  \\\n",
       "Symbol                                                               ...   \n",
       "TSPAN6       6     8    10     6     4     2     5     3     5    3  ...   \n",
       "TNMD         0     1     0     0     1     1     0     0     0    0  ...   \n",
       "DPM1        20    35    14    22    17     9    30    22    22   16  ...   \n",
       "SCYL3       18    10     6    18     9     9    11    18     8    6  ...   \n",
       "C1orf112     1     2     1     1     2     0     0     1     0    0  ...   \n",
       "\n",
       "          X_75  X_76  X_77  X_78  X_79  X_7  X_80  X_81  X_8  X_9  \n",
       "Symbol                                                             \n",
       "TSPAN6       5     7     6     9     3    3     7     5    4    1  \n",
       "TNMD         0     2     0     0     0    0     0     0    0    0  \n",
       "DPM1        34    22    35    20    23   19    40    12    6   34  \n",
       "SCYL3       11    13    11    12    12    9     5     8    5   10  \n",
       "C1orf112     2     0     0     1     1    0     0     1    2    0  \n",
       "\n",
       "[5 rows x 76 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Count table 로드\n",
    "counts = pd.read_csv('../1_GSE167186/GSE167186_counts.csv', index_col=0)\n",
    "print(counts.shape)\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "497cd33e-4046-47e3-81be-744e72099bce",
   "metadata": {},
   "source": [
    "다음으로 **Metadata를 확인**해봅니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5accb13b-e85a-41c6-a896-3fa49a0dc0f9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X_10</th>\n",
       "      <th>X_11</th>\n",
       "      <th>X_13</th>\n",
       "      <th>X_14</th>\n",
       "      <th>X_15</th>\n",
       "      <th>X_16</th>\n",
       "      <th>X_17</th>\n",
       "      <th>X_18</th>\n",
       "      <th>X_19</th>\n",
       "      <th>X_1</th>\n",
       "      <th>...</th>\n",
       "      <th>X_75</th>\n",
       "      <th>X_76</th>\n",
       "      <th>X_77</th>\n",
       "      <th>X_78</th>\n",
       "      <th>X_79</th>\n",
       "      <th>X_7</th>\n",
       "      <th>X_80</th>\n",
       "      <th>X_81</th>\n",
       "      <th>X_8</th>\n",
       "      <th>X_9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_title</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>!Sample_geo_accession</th>\n",
       "      <td>GSM5098661</td>\n",
       "      <td>GSM5098662</td>\n",
       "      <td>GSM5098663</td>\n",
       "      <td>GSM5098664</td>\n",
       "      <td>GSM5098665</td>\n",
       "      <td>GSM5098666</td>\n",
       "      <td>GSM5098667</td>\n",
       "      <td>GSM5098668</td>\n",
       "      <td>GSM5098669</td>\n",
       "      <td>GSM5098670</td>\n",
       "      <td>...</td>\n",
       "      <td>GSM5098727</td>\n",
       "      <td>GSM5098728</td>\n",
       "      <td>GSM5098729</td>\n",
       "      <td>GSM5098730</td>\n",
       "      <td>GSM5098731</td>\n",
       "      <td>GSM5098732</td>\n",
       "      <td>GSM5098733</td>\n",
       "      <td>GSM5098734</td>\n",
       "      <td>GSM5098735</td>\n",
       "      <td>GSM5098736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_status</th>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>...</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "      <td>Public on Mar 31 2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_submission_date</th>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>...</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "      <td>Feb 21 2021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_last_update_date</th>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>...</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "      <td>Sep 14 2023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_type</th>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>...</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "      <td>SRA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           X_10                   X_11  \\\n",
       "!Sample_title                                                            \n",
       "!Sample_geo_accession                GSM5098661             GSM5098662   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023   \n",
       "!Sample_type                                SRA                    SRA   \n",
       "\n",
       "                                           X_13                   X_14  \\\n",
       "!Sample_title                                                            \n",
       "!Sample_geo_accession                GSM5098663             GSM5098664   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023   \n",
       "!Sample_type                                SRA                    SRA   \n",
       "\n",
       "                                           X_15                   X_16  \\\n",
       "!Sample_title                                                            \n",
       "!Sample_geo_accession                GSM5098665             GSM5098666   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023   \n",
       "!Sample_type                                SRA                    SRA   \n",
       "\n",
       "                                           X_17                   X_18  \\\n",
       "!Sample_title                                                            \n",
       "!Sample_geo_accession                GSM5098667             GSM5098668   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023   \n",
       "!Sample_type                                SRA                    SRA   \n",
       "\n",
       "                                           X_19                    X_1  ...  \\\n",
       "!Sample_title                                                           ...   \n",
       "!Sample_geo_accession                GSM5098669             GSM5098670  ...   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022  ...   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021  ...   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023  ...   \n",
       "!Sample_type                                SRA                    SRA  ...   \n",
       "\n",
       "                                           X_75                   X_76  \\\n",
       "!Sample_title                                                            \n",
       "!Sample_geo_accession                GSM5098727             GSM5098728   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023   \n",
       "!Sample_type                                SRA                    SRA   \n",
       "\n",
       "                                           X_77                   X_78  \\\n",
       "!Sample_title                                                            \n",
       "!Sample_geo_accession                GSM5098729             GSM5098730   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023   \n",
       "!Sample_type                                SRA                    SRA   \n",
       "\n",
       "                                           X_79                    X_7  \\\n",
       "!Sample_title                                                            \n",
       "!Sample_geo_accession                GSM5098731             GSM5098732   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023   \n",
       "!Sample_type                                SRA                    SRA   \n",
       "\n",
       "                                           X_80                   X_81  \\\n",
       "!Sample_title                                                            \n",
       "!Sample_geo_accession                GSM5098733             GSM5098734   \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022   \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021   \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023   \n",
       "!Sample_type                                SRA                    SRA   \n",
       "\n",
       "                                            X_8                    X_9  \n",
       "!Sample_title                                                           \n",
       "!Sample_geo_accession                GSM5098735             GSM5098736  \n",
       "!Sample_status            Public on Mar 31 2022  Public on Mar 31 2022  \n",
       "!Sample_submission_date             Feb 21 2021            Feb 21 2021  \n",
       "!Sample_last_update_date            Sep 14 2023            Sep 14 2023  \n",
       "!Sample_type                                SRA                    SRA  \n",
       "\n",
       "[5 rows x 76 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Metadata 로드\n",
    "meta = pd.read_table('../1_GSE167186/GSE167186-GPL20301_series_matrix.txt', sep='\\t', skiprows=30, index_col=0)\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8a7c2f3-5968-4b9e-9c34-b582081a0169",
   "metadata": {},
   "source": [
    "`!Sample_characteristics_ch1` 행에 **환자 그룹의 특징이 저장**되어 있습니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5762b223-9e15-4053-9063-5a21b23a94c2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['!Sample_geo_accession', '!Sample_status', '!Sample_submission_date',\n",
      "       '!Sample_last_update_date', '!Sample_type', '!Sample_channel_count',\n",
      "       '!Sample_source_name_ch1', '!Sample_organism_ch1',\n",
      "       '!Sample_characteristics_ch1', '!Sample_characteristics_ch1',\n",
      "       '!Sample_characteristics_ch1', '!Sample_characteristics_ch1',\n",
      "       '!Sample_characteristics_ch1', '!Sample_characteristics_ch1',\n",
      "       '!Sample_characteristics_ch1', '!Sample_characteristics_ch1',\n",
      "       '!Sample_characteristics_ch1', '!Sample_molecule_ch1',\n",
      "       '!Sample_extract_protocol_ch1', '!Sample_extract_protocol_ch1',\n",
      "       '!Sample_extract_protocol_ch1', '!Sample_extract_protocol_ch1',\n",
      "       '!Sample_taxid_ch1', '!Sample_description', '!Sample_description',\n",
      "       '!Sample_description', '!Sample_data_processing',\n",
      "       '!Sample_data_processing', '!Sample_data_processing',\n",
      "       '!Sample_data_processing', '!Sample_data_processing',\n",
      "       '!Sample_platform_id', '!Sample_contact_name', '!Sample_contact_email',\n",
      "       '!Sample_contact_laboratory', '!Sample_contact_institute',\n",
      "       '!Sample_contact_address', '!Sample_contact_city',\n",
      "       '!Sample_contact_state', '!Sample_contact_zip/postal_code',\n",
      "       '!Sample_contact_country', '!Sample_data_row_count',\n",
      "       '!Sample_instrument_model', '!Sample_library_selection',\n",
      "       '!Sample_library_source', '!Sample_library_strategy',\n",
      "       '!Sample_relation', '!Sample_relation', '!Sample_supplementary_file_1',\n",
      "       '!series_matrix_table_begin', 'ID_REF', '!series_matrix_table_end'],\n",
      "      dtype='object', name='!Sample_title')\n"
     ]
    }
   ],
   "source": [
    "print(meta.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a65b13a0-3e3f-4f8c-9d12-c05d68868217",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X_10</th>\n",
       "      <th>X_11</th>\n",
       "      <th>X_13</th>\n",
       "      <th>X_14</th>\n",
       "      <th>X_15</th>\n",
       "      <th>X_16</th>\n",
       "      <th>X_17</th>\n",
       "      <th>X_18</th>\n",
       "      <th>X_19</th>\n",
       "      <th>X_1</th>\n",
       "      <th>...</th>\n",
       "      <th>X_75</th>\n",
       "      <th>X_76</th>\n",
       "      <th>X_77</th>\n",
       "      <th>X_78</th>\n",
       "      <th>X_79</th>\n",
       "      <th>X_7</th>\n",
       "      <th>X_80</th>\n",
       "      <th>X_81</th>\n",
       "      <th>X_8</th>\n",
       "      <th>X_9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_title</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>group: Sarcopenia</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Sarcopenia</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Sarcopenia</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Sarcopenia</td>\n",
       "      <td>group: Sarcopenia</td>\n",
       "      <td>...</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Sarcopenia</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: UNCLASSIFIED</td>\n",
       "      <td>group: Sarcopenia</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Sarcopenia</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "      <td>group: Old Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "      <td>tissue: Skeletal muscle</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>age: 68</td>\n",
       "      <td>age: 87</td>\n",
       "      <td>age: 83</td>\n",
       "      <td>age: 77</td>\n",
       "      <td>age: 82</td>\n",
       "      <td>age: 73</td>\n",
       "      <td>age: 77.9</td>\n",
       "      <td>age: 64</td>\n",
       "      <td>age: 77</td>\n",
       "      <td>age: 67</td>\n",
       "      <td>...</td>\n",
       "      <td>age: 66</td>\n",
       "      <td>age: 70</td>\n",
       "      <td>age: 71</td>\n",
       "      <td>age: 66</td>\n",
       "      <td>age: 71</td>\n",
       "      <td>age: 73</td>\n",
       "      <td>age: 71</td>\n",
       "      <td>age: 67</td>\n",
       "      <td>age: 76</td>\n",
       "      <td>age: 66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>6 min walk test (m/s): 1.01010101</td>\n",
       "      <td>6 min walk test (m/s): 1.08892922</td>\n",
       "      <td>6 min walk test (m/s): 1.425178147</td>\n",
       "      <td>6 min walk test (m/s): 0.918836141</td>\n",
       "      <td>biodex (kg): 127.9</td>\n",
       "      <td>biodex (kg): 202.7</td>\n",
       "      <td>6 min walk test (m/s): 0.886262925</td>\n",
       "      <td>6 min walk test (m/s): 1.086956522</td>\n",
       "      <td>6 min walk test (m/s): 1.041666667</td>\n",
       "      <td>biodex (kg): 148.7</td>\n",
       "      <td>...</td>\n",
       "      <td>6 min walk test (m/s): 1.366742597</td>\n",
       "      <td>6 min walk test (m/s): 1.363636364</td>\n",
       "      <td>6 min walk test (m/s): 0.896860987</td>\n",
       "      <td>6 min walk test (m/s): 1.388888889</td>\n",
       "      <td>6 min walk test (m/s): 1.098901099</td>\n",
       "      <td>6 min walk test (m/s): 1.229508197</td>\n",
       "      <td>6 min walk test (m/s): 1.054481547</td>\n",
       "      <td>6 min walk test (m/s): 0.930232558</td>\n",
       "      <td>6 min walk test (m/s): 1.096892139</td>\n",
       "      <td>6 min walk test (m/s): 0.805369128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>biodex (kg): 166.2</td>\n",
       "      <td>biodex (kg): 136.4</td>\n",
       "      <td>biodex (kg): 143.1</td>\n",
       "      <td>biodex (kg): 222.1</td>\n",
       "      <td>time up and go (s): 7.16</td>\n",
       "      <td>time up and go (s): 8.27</td>\n",
       "      <td>biodex (kg): 171.7</td>\n",
       "      <td>biodex (kg): 203.9</td>\n",
       "      <td>biodex (kg): 143.8</td>\n",
       "      <td>time up and go (s): 7.85</td>\n",
       "      <td>...</td>\n",
       "      <td>biodex (kg): 236.1</td>\n",
       "      <td>biodex (kg): 244.1</td>\n",
       "      <td>biodex (kg): 106.5</td>\n",
       "      <td>biodex (kg): 239.6</td>\n",
       "      <td>biodex (kg): 195.3</td>\n",
       "      <td>biodex (kg): 155</td>\n",
       "      <td>biodex (kg): 215</td>\n",
       "      <td>biodex (kg): 184.4</td>\n",
       "      <td>biodex (kg): 144.5</td>\n",
       "      <td>biodex (kg): 208.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>time up and go (s): 7.15</td>\n",
       "      <td>time up and go (s): 7.82</td>\n",
       "      <td>time up and go (s): 7.97</td>\n",
       "      <td>time up and go (s): 6.56</td>\n",
       "      <td>sppb: 12</td>\n",
       "      <td>sppb: 12</td>\n",
       "      <td>time up and go (s): 7.21</td>\n",
       "      <td>time up and go (s): 6.83</td>\n",
       "      <td>time up and go (s): 8.37</td>\n",
       "      <td>grip strength (kg): 41.3</td>\n",
       "      <td>...</td>\n",
       "      <td>time up and go (s): 6.1</td>\n",
       "      <td>time up and go (s): 7.94</td>\n",
       "      <td>time up and go (s): 8.96</td>\n",
       "      <td>time up and go (s): 5.97</td>\n",
       "      <td>time up and go (s): 7.01</td>\n",
       "      <td>time up and go (s): 7.59</td>\n",
       "      <td>time up and go (s): 7.91</td>\n",
       "      <td>time up and go (s): 7.27</td>\n",
       "      <td>time up and go (s): 8.2</td>\n",
       "      <td>time up and go (s): 7.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>sppb: 12</td>\n",
       "      <td>sppb: 11</td>\n",
       "      <td>sppb: 10</td>\n",
       "      <td>sppb: 11</td>\n",
       "      <td>grip strength (kg): 35.1</td>\n",
       "      <td>grip strength (kg): 49.1</td>\n",
       "      <td>sppb: 9</td>\n",
       "      <td>sppb: 12</td>\n",
       "      <td>sppb: 7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>sppb: 12</td>\n",
       "      <td>sppb: 12</td>\n",
       "      <td>sppb: 8</td>\n",
       "      <td>sppb: 11</td>\n",
       "      <td>sppb: 12</td>\n",
       "      <td>sppb: 11</td>\n",
       "      <td>sppb: 11</td>\n",
       "      <td>sppb: 11</td>\n",
       "      <td>sppb: 10</td>\n",
       "      <td>sppb: 10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>grip strength (kg): 41.8</td>\n",
       "      <td>grip strength (kg): 24.2</td>\n",
       "      <td>grip strength (kg): 40</td>\n",
       "      <td>grip strength (kg): 35.5</td>\n",
       "      <td>leg press (kg): 94.5</td>\n",
       "      <td>leg press (kg): 145.5</td>\n",
       "      <td>grip strength (kg): 45.5</td>\n",
       "      <td>grip strength (kg): 42.7</td>\n",
       "      <td>grip strength (kg): 33.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>leg press (kg): 157.5</td>\n",
       "      <td>grip strength (kg): 47.6</td>\n",
       "      <td>grip strength (kg): 28.2</td>\n",
       "      <td>grip strength (kg): 42.2</td>\n",
       "      <td>leg press (kg): 139.5</td>\n",
       "      <td>grip strength (kg): 34.3</td>\n",
       "      <td>leg press (kg): 130.5</td>\n",
       "      <td>grip strength (kg): 40.1</td>\n",
       "      <td>grip strength (kg): 41.5</td>\n",
       "      <td>grip strength (kg): 42.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "      <td>leg press (kg): 130.5</td>\n",
       "      <td>leg press (kg): 129.55</td>\n",
       "      <td>leg press (kg): 127.27</td>\n",
       "      <td>leg press (kg): 229.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>leg press (kg): 87.5</td>\n",
       "      <td>leg press (kg): 112.5</td>\n",
       "      <td>leg press (kg): 112.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>leg press (kg): 166</td>\n",
       "      <td>leg press (kg): 63.5</td>\n",
       "      <td>leg press (kg): 166.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>leg press (kg): 127.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>leg press (kg): 130.5</td>\n",
       "      <td>leg press (kg): 94.5</td>\n",
       "      <td>leg press (kg): 147.73</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          X_10  \\\n",
       "!Sample_title                                                    \n",
       "!Sample_characteristics_ch1                  group: Sarcopenia   \n",
       "!Sample_characteristics_ch1            tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                            age: 68   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.01010101   \n",
       "!Sample_characteristics_ch1                 biodex (kg): 166.2   \n",
       "!Sample_characteristics_ch1           time up and go (s): 7.15   \n",
       "!Sample_characteristics_ch1                           sppb: 12   \n",
       "!Sample_characteristics_ch1           grip strength (kg): 41.8   \n",
       "!Sample_characteristics_ch1              leg press (kg): 130.5   \n",
       "\n",
       "                                                          X_11  \\\n",
       "!Sample_title                                                    \n",
       "!Sample_characteristics_ch1                 group: Old Healthy   \n",
       "!Sample_characteristics_ch1            tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                            age: 87   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.08892922   \n",
       "!Sample_characteristics_ch1                 biodex (kg): 136.4   \n",
       "!Sample_characteristics_ch1           time up and go (s): 7.82   \n",
       "!Sample_characteristics_ch1                           sppb: 11   \n",
       "!Sample_characteristics_ch1           grip strength (kg): 24.2   \n",
       "!Sample_characteristics_ch1             leg press (kg): 129.55   \n",
       "\n",
       "                                                           X_13  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                   group: Sarcopenia   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 83   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.425178147   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 143.1   \n",
       "!Sample_characteristics_ch1            time up and go (s): 7.97   \n",
       "!Sample_characteristics_ch1                            sppb: 10   \n",
       "!Sample_characteristics_ch1              grip strength (kg): 40   \n",
       "!Sample_characteristics_ch1              leg press (kg): 127.27   \n",
       "\n",
       "                                                           X_14  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                  group: Old Healthy   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 77   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 0.918836141   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 222.1   \n",
       "!Sample_characteristics_ch1            time up and go (s): 6.56   \n",
       "!Sample_characteristics_ch1                            sppb: 11   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 35.5   \n",
       "!Sample_characteristics_ch1               leg press (kg): 229.5   \n",
       "\n",
       "                                                 X_15  \\\n",
       "!Sample_title                                           \n",
       "!Sample_characteristics_ch1        group: Old Healthy   \n",
       "!Sample_characteristics_ch1   tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                   age: 82   \n",
       "!Sample_characteristics_ch1        biodex (kg): 127.9   \n",
       "!Sample_characteristics_ch1  time up and go (s): 7.16   \n",
       "!Sample_characteristics_ch1                  sppb: 12   \n",
       "!Sample_characteristics_ch1  grip strength (kg): 35.1   \n",
       "!Sample_characteristics_ch1      leg press (kg): 94.5   \n",
       "!Sample_characteristics_ch1                       NaN   \n",
       "\n",
       "                                                 X_16  \\\n",
       "!Sample_title                                           \n",
       "!Sample_characteristics_ch1        group: Old Healthy   \n",
       "!Sample_characteristics_ch1   tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                   age: 73   \n",
       "!Sample_characteristics_ch1        biodex (kg): 202.7   \n",
       "!Sample_characteristics_ch1  time up and go (s): 8.27   \n",
       "!Sample_characteristics_ch1                  sppb: 12   \n",
       "!Sample_characteristics_ch1  grip strength (kg): 49.1   \n",
       "!Sample_characteristics_ch1     leg press (kg): 145.5   \n",
       "!Sample_characteristics_ch1                       NaN   \n",
       "\n",
       "                                                           X_17  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                   group: Sarcopenia   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                           age: 77.9   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 0.886262925   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 171.7   \n",
       "!Sample_characteristics_ch1            time up and go (s): 7.21   \n",
       "!Sample_characteristics_ch1                             sppb: 9   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 45.5   \n",
       "!Sample_characteristics_ch1                leg press (kg): 87.5   \n",
       "\n",
       "                                                           X_18  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                  group: Old Healthy   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 64   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.086956522   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 203.9   \n",
       "!Sample_characteristics_ch1            time up and go (s): 6.83   \n",
       "!Sample_characteristics_ch1                            sppb: 12   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 42.7   \n",
       "!Sample_characteristics_ch1               leg press (kg): 112.5   \n",
       "\n",
       "                                                           X_19  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                   group: Sarcopenia   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 77   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.041666667   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 143.8   \n",
       "!Sample_characteristics_ch1            time up and go (s): 8.37   \n",
       "!Sample_characteristics_ch1                             sppb: 7   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 33.7   \n",
       "!Sample_characteristics_ch1               leg press (kg): 112.5   \n",
       "\n",
       "                                                  X_1  ...  \\\n",
       "!Sample_title                                          ...   \n",
       "!Sample_characteristics_ch1         group: Sarcopenia  ...   \n",
       "!Sample_characteristics_ch1   tissue: Skeletal muscle  ...   \n",
       "!Sample_characteristics_ch1                   age: 67  ...   \n",
       "!Sample_characteristics_ch1        biodex (kg): 148.7  ...   \n",
       "!Sample_characteristics_ch1  time up and go (s): 7.85  ...   \n",
       "!Sample_characteristics_ch1  grip strength (kg): 41.3  ...   \n",
       "!Sample_characteristics_ch1                       NaN  ...   \n",
       "!Sample_characteristics_ch1                       NaN  ...   \n",
       "!Sample_characteristics_ch1                       NaN  ...   \n",
       "\n",
       "                                                           X_75  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                  group: Old Healthy   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 66   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.366742597   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 236.1   \n",
       "!Sample_characteristics_ch1             time up and go (s): 6.1   \n",
       "!Sample_characteristics_ch1                            sppb: 12   \n",
       "!Sample_characteristics_ch1               leg press (kg): 157.5   \n",
       "!Sample_characteristics_ch1                                 NaN   \n",
       "\n",
       "                                                           X_76  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                  group: Old Healthy   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 70   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.363636364   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 244.1   \n",
       "!Sample_characteristics_ch1            time up and go (s): 7.94   \n",
       "!Sample_characteristics_ch1                            sppb: 12   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 47.6   \n",
       "!Sample_characteristics_ch1                 leg press (kg): 166   \n",
       "\n",
       "                                                           X_77  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                   group: Sarcopenia   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 71   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 0.896860987   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 106.5   \n",
       "!Sample_characteristics_ch1            time up and go (s): 8.96   \n",
       "!Sample_characteristics_ch1                             sppb: 8   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 28.2   \n",
       "!Sample_characteristics_ch1                leg press (kg): 63.5   \n",
       "\n",
       "                                                           X_78  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                  group: Old Healthy   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 66   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.388888889   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 239.6   \n",
       "!Sample_characteristics_ch1            time up and go (s): 5.97   \n",
       "!Sample_characteristics_ch1                            sppb: 11   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 42.2   \n",
       "!Sample_characteristics_ch1               leg press (kg): 166.5   \n",
       "\n",
       "                                                           X_79  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                 group: UNCLASSIFIED   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 71   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.098901099   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 195.3   \n",
       "!Sample_characteristics_ch1            time up and go (s): 7.01   \n",
       "!Sample_characteristics_ch1                            sppb: 12   \n",
       "!Sample_characteristics_ch1               leg press (kg): 139.5   \n",
       "!Sample_characteristics_ch1                                 NaN   \n",
       "\n",
       "                                                            X_7  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                   group: Sarcopenia   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 73   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.229508197   \n",
       "!Sample_characteristics_ch1                    biodex (kg): 155   \n",
       "!Sample_characteristics_ch1            time up and go (s): 7.59   \n",
       "!Sample_characteristics_ch1                            sppb: 11   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 34.3   \n",
       "!Sample_characteristics_ch1              leg press (kg): 127.27   \n",
       "\n",
       "                                                           X_80  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                  group: Old Healthy   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 71   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.054481547   \n",
       "!Sample_characteristics_ch1                    biodex (kg): 215   \n",
       "!Sample_characteristics_ch1            time up and go (s): 7.91   \n",
       "!Sample_characteristics_ch1                            sppb: 11   \n",
       "!Sample_characteristics_ch1               leg press (kg): 130.5   \n",
       "!Sample_characteristics_ch1                                 NaN   \n",
       "\n",
       "                                                           X_81  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                   group: Sarcopenia   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 67   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 0.930232558   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 184.4   \n",
       "!Sample_characteristics_ch1            time up and go (s): 7.27   \n",
       "!Sample_characteristics_ch1                            sppb: 11   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 40.1   \n",
       "!Sample_characteristics_ch1               leg press (kg): 130.5   \n",
       "\n",
       "                                                            X_8  \\\n",
       "!Sample_title                                                     \n",
       "!Sample_characteristics_ch1                  group: Old Healthy   \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle   \n",
       "!Sample_characteristics_ch1                             age: 76   \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 1.096892139   \n",
       "!Sample_characteristics_ch1                  biodex (kg): 144.5   \n",
       "!Sample_characteristics_ch1             time up and go (s): 8.2   \n",
       "!Sample_characteristics_ch1                            sppb: 10   \n",
       "!Sample_characteristics_ch1            grip strength (kg): 41.5   \n",
       "!Sample_characteristics_ch1                leg press (kg): 94.5   \n",
       "\n",
       "                                                            X_9  \n",
       "!Sample_title                                                    \n",
       "!Sample_characteristics_ch1                  group: Old Healthy  \n",
       "!Sample_characteristics_ch1             tissue: Skeletal muscle  \n",
       "!Sample_characteristics_ch1                             age: 66  \n",
       "!Sample_characteristics_ch1  6 min walk test (m/s): 0.805369128  \n",
       "!Sample_characteristics_ch1                  biodex (kg): 208.8  \n",
       "!Sample_characteristics_ch1            time up and go (s): 7.82  \n",
       "!Sample_characteristics_ch1                            sppb: 10  \n",
       "!Sample_characteristics_ch1            grip strength (kg): 42.2  \n",
       "!Sample_characteristics_ch1              leg press (kg): 147.73  \n",
       "\n",
       "[9 rows x 76 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.loc['!Sample_characteristics_ch1']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbd57d5d-93b4-45f5-a428-6833f56f20e8",
   "metadata": {},
   "source": [
    "`!Sample_characteristics_ch1` 행\b에서 **그룹 정보가 들어간 행**을 사용할겁니다.  \n",
    "다음에서 개수를 확인해봅니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2c17d9b3-c63d-43e8-9b3e-a32d452cc66e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "group: Old Healthy      29\n",
       "group: Sarcopenia       24\n",
       "group: Young Healthy    19\n",
       "group: UNCLASSIFIED      4\n",
       "Name: !Sample_characteristics_ch1, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.iloc[8].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dbcb784-ff68-4630-8c3d-fe3849c4cda3",
   "metadata": {},
   "source": [
    "# Extract dataframe for EDA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6ceea6e0-1e41-40b2-894f-db61c8047ba5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 76 entries, X_10 to X_9\n",
      "Data columns (total 9 columns):\n",
      " #   Column                       Non-Null Count  Dtype \n",
      "---  ------                       --------------  ----- \n",
      " 0   !Sample_characteristics_ch1  76 non-null     object\n",
      " 1   !Sample_characteristics_ch1  76 non-null     object\n",
      " 2   !Sample_characteristics_ch1  76 non-null     object\n",
      " 3   !Sample_characteristics_ch1  76 non-null     object\n",
      " 4   !Sample_characteristics_ch1  76 non-null     object\n",
      " 5   !Sample_characteristics_ch1  76 non-null     object\n",
      " 6   !Sample_characteristics_ch1  74 non-null     object\n",
      " 7   !Sample_characteristics_ch1  74 non-null     object\n",
      " 8   !Sample_characteristics_ch1  67 non-null     object\n",
      "dtypes: object(9)\n",
      "memory usage: 8.0+ KB\n"
     ]
    }
   ],
   "source": [
    "meta.loc['!Sample_characteristics_ch1'].T.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b403bce0-4f24-4159-971e-c7ea752fbc96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# nan 값 처리\n",
    "meta.loc['!Sample_characteristics_ch1'] = meta.loc['!Sample_characteristics_ch1'].fillna(': ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2488aef6-e233-4e2b-831e-90b9c61ffe77",
   "metadata": {},
   "outputs": [],
   "source": [
    "eda = {}\n",
    "for i in range(0,9):\n",
    "    key = meta.loc['!Sample_characteristics_ch1'].iloc[i][0].split(': ')[0]\n",
    "    value = meta.loc['!Sample_characteristics_ch1'].iloc[i]\n",
    "    value = value.apply(lambda x:x.split(': ')[-1])\n",
    "    eda[key] = value\n",
    "\n",
    "eda = pd.DataFrame(eda)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d3cc5d4f-001e-4976-8093-9e890bdd70c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>tissue</th>\n",
       "      <th>age</th>\n",
       "      <th>6 min walk test (m/s)</th>\n",
       "      <th>biodex (kg)</th>\n",
       "      <th>time up and go (s)</th>\n",
       "      <th>sppb</th>\n",
       "      <th>grip strength (kg)</th>\n",
       "      <th>leg press (kg)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X_10</th>\n",
       "      <td>Sarcopenia</td>\n",
       "      <td>Skeletal muscle</td>\n",
       "      <td>68</td>\n",
       "      <td>1.01010101</td>\n",
       "      <td>166.2</td>\n",
       "      <td>7.15</td>\n",
       "      <td>12</td>\n",
       "      <td>41.8</td>\n",
       "      <td>130.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_11</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>Skeletal muscle</td>\n",
       "      <td>87</td>\n",
       "      <td>1.08892922</td>\n",
       "      <td>136.4</td>\n",
       "      <td>7.82</td>\n",
       "      <td>11</td>\n",
       "      <td>24.2</td>\n",
       "      <td>129.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_13</th>\n",
       "      <td>Sarcopenia</td>\n",
       "      <td>Skeletal muscle</td>\n",
       "      <td>83</td>\n",
       "      <td>1.425178147</td>\n",
       "      <td>143.1</td>\n",
       "      <td>7.97</td>\n",
       "      <td>10</td>\n",
       "      <td>40</td>\n",
       "      <td>127.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_14</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>Skeletal muscle</td>\n",
       "      <td>77</td>\n",
       "      <td>0.918836141</td>\n",
       "      <td>222.1</td>\n",
       "      <td>6.56</td>\n",
       "      <td>11</td>\n",
       "      <td>35.5</td>\n",
       "      <td>229.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_15</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>Skeletal muscle</td>\n",
       "      <td>82</td>\n",
       "      <td>127.9</td>\n",
       "      <td>7.16</td>\n",
       "      <td>12</td>\n",
       "      <td>35.1</td>\n",
       "      <td>94.5</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            group           tissue age 6 min walk test (m/s) biodex (kg)  \\\n",
       "X_10   Sarcopenia  Skeletal muscle  68            1.01010101       166.2   \n",
       "X_11  Old Healthy  Skeletal muscle  87            1.08892922       136.4   \n",
       "X_13   Sarcopenia  Skeletal muscle  83           1.425178147       143.1   \n",
       "X_14  Old Healthy  Skeletal muscle  77           0.918836141       222.1   \n",
       "X_15  Old Healthy  Skeletal muscle  82                 127.9        7.16   \n",
       "\n",
       "     time up and go (s)  sppb grip strength (kg) leg press (kg)  \n",
       "X_10               7.15    12               41.8          130.5  \n",
       "X_11               7.82    11               24.2         129.55  \n",
       "X_13               7.97    10                 40         127.27  \n",
       "X_14               6.56    11               35.5          229.5  \n",
       "X_15                 12  35.1               94.5                 "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eda.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "65a288cf-7d54-4df2-a034-9a15696db9c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# os.chdir('/Users/soheepark/03-GEO근감소/Data/')\n",
    "# eda.to_csv('./Results/GSE167186_eda.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51f70c5f-9247-4f31-b566-0dd9d3296684",
   "metadata": {},
   "source": [
    "# Preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "091d5f9f-11c2-461f-88bf-10a0a80dcff1",
   "metadata": {},
   "source": [
    "## Preperation of metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d0f8f56-d6d2-48d9-bd9d-cdedac911625",
   "metadata": {},
   "source": [
    "다음은 Metadata 정보를 분석에 바로 활용할 수 있도록 전처리하는 과정입니다.  \n",
    "1. dataframe으로 만듭니다.\n",
    "2. index 기준 오름차순으로 정렬합니다.\n",
    "3. 불필요한 문자열을 삭제합니다.\n",
    "4. 분석에 필요한 그룹만 남겨둡니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5c416ef1-2fe6-4a3c-bc8d-4e438a197d38",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['group: Sarcopenia' 'group: Old Healthy' 'group: UNCLASSIFIED'\n",
      " 'group: Young Healthy']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>!Sample_characteristics_ch1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X_10</th>\n",
       "      <td>group: Sarcopenia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_11</th>\n",
       "      <td>group: Old Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_13</th>\n",
       "      <td>group: Sarcopenia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_14</th>\n",
       "      <td>group: Old Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_15</th>\n",
       "      <td>group: Old Healthy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     !Sample_characteristics_ch1\n",
       "X_10           group: Sarcopenia\n",
       "X_11          group: Old Healthy\n",
       "X_13           group: Sarcopenia\n",
       "X_14          group: Old Healthy\n",
       "X_15          group: Old Healthy"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# group을 나타내는 Metadata를 dataframe으로 만들기\n",
    "metadata = pd.DataFrame(meta.iloc[8])\n",
    "print(metadata[metadata.columns[0]].unique())\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7b5e6fd8-ee05-4b77-8619-9c97655de4cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "metaname1 = {metadata[metadata.columns[0]].unique()[0]:'Old Sarcopenia',\n",
    "            metadata[metadata.columns[0]].unique()[1]:'Old Healthy',\n",
    "            metadata[metadata.columns[0]].unique()[2]:'Unclassified',\n",
    "            metadata[metadata.columns[0]].unique()[3]:'Young Healthy'}\n",
    "metaname2 = {metadata[metadata.columns[0]].unique()[0]:'OS',\n",
    "            metadata[metadata.columns[0]].unique()[1]:'OH',\n",
    "            metadata[metadata.columns[0]].unique()[2]:'Un',\n",
    "            metadata[metadata.columns[0]].unique()[3]:'YH'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8042470d-8797-4a80-bc6f-1382321253c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition_</th>\n",
       "      <th>Condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X_10</th>\n",
       "      <td>Old Sarcopenia</td>\n",
       "      <td>OS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_11</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>OH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_13</th>\n",
       "      <td>Old Sarcopenia</td>\n",
       "      <td>OS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_14</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>OH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_15</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>OH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Condition_ Condition\n",
       "X_10  Old Sarcopenia        OS\n",
       "X_11     Old Healthy        OH\n",
       "X_13  Old Sarcopenia        OS\n",
       "X_14     Old Healthy        OH\n",
       "X_15     Old Healthy        OH"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['Condition_'] = metadata[metadata.columns[0]].replace(metaname1)\n",
    "metadata['Condition'] = metadata[metadata.columns[0]].replace(metaname2)\n",
    "del metadata[metadata.columns[0]]\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "51c0f400-b5b9-417a-ab4a-362b9f1a6822",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Condition_</th>\n",
       "      <th>Condition</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Temp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X_10</th>\n",
       "      <td>Old Sarcopenia</td>\n",
       "      <td>OS</td>\n",
       "      <td>X_10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_11</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>OH</td>\n",
       "      <td>X_11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_13</th>\n",
       "      <td>Old Sarcopenia</td>\n",
       "      <td>OS</td>\n",
       "      <td>X_13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_14</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>OH</td>\n",
       "      <td>X_14</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_15</th>\n",
       "      <td>Old Healthy</td>\n",
       "      <td>OH</td>\n",
       "      <td>X_15</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Condition_ Condition Sample Temp\n",
       "X_10  Old Sarcopenia        OS   X_10   10\n",
       "X_11     Old Healthy        OH   X_11   11\n",
       "X_13  Old Sarcopenia        OS   X_13   13\n",
       "X_14     Old Healthy        OH   X_14   14\n",
       "X_15     Old Healthy        OH   X_15   15"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['Sample'] = metadata.index\n",
    "metadata['Temp'] = metadata['Sample'].apply(lambda x:x.split('_')[-1])\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "793d5f18-2218-4ea8-8854-72c7252eb78f",
   "metadata": {},
   "source": [
    "## Preperation of count data\n",
    "다음은 Count data를 분석에 바로 활용할 수 있도록 전처리하는 과정입니다.\n",
    "1. 분석하지 않는 Unclassified 그룹을 제외합니다.\n",
    "2. 행에 샘플, 열에 유전자 이름이 나오도록 transpose 합니다."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0e22a99c-85c6-42dc-be21-0bd17f6f35fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X_10</th>\n",
       "      <th>X_11</th>\n",
       "      <th>X_13</th>\n",
       "      <th>X_14</th>\n",
       "      <th>X_15</th>\n",
       "      <th>X_16</th>\n",
       "      <th>X_17</th>\n",
       "      <th>X_18</th>\n",
       "      <th>X_19</th>\n",
       "      <th>X_1</th>\n",
       "      <th>...</th>\n",
       "      <th>X_75</th>\n",
       "      <th>X_76</th>\n",
       "      <th>X_77</th>\n",
       "      <th>X_78</th>\n",
       "      <th>X_79</th>\n",
       "      <th>X_7</th>\n",
       "      <th>X_80</th>\n",
       "      <th>X_81</th>\n",
       "      <th>X_8</th>\n",
       "      <th>X_9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>20</td>\n",
       "      <td>35</td>\n",
       "      <td>14</td>\n",
       "      <td>22</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>34</td>\n",
       "      <td>22</td>\n",
       "      <td>35</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>12</td>\n",
       "      <td>6</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>18</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>18</td>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>11</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 76 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          X_10  X_11  X_13  X_14  X_15  X_16  X_17  X_18  X_19  X_1  ...  \\\n",
       "Symbol                                                               ...   \n",
       "TSPAN6       6     8    10     6     4     2     5     3     5    3  ...   \n",
       "TNMD         0     1     0     0     1     1     0     0     0    0  ...   \n",
       "DPM1        20    35    14    22    17     9    30    22    22   16  ...   \n",
       "SCYL3       18    10     6    18     9     9    11    18     8    6  ...   \n",
       "C1orf112     1     2     1     1     2     0     0     1     0    0  ...   \n",
       "\n",
       "          X_75  X_76  X_77  X_78  X_79  X_7  X_80  X_81  X_8  X_9  \n",
       "Symbol                                                             \n",
       "TSPAN6       5     7     6     9     3    3     7     5    4    1  \n",
       "TNMD         0     2     0     0     0    0     0     0    0    0  \n",
       "DPM1        34    22    35    20    23   19    40    12    6   34  \n",
       "SCYL3       11    13    11    12    12    9     5     8    5   10  \n",
       "C1orf112     2     0     0     1     1    0     0     1    2    0  \n",
       "\n",
       "[5 rows x 76 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# UNCLASSIFIED 제외한 테이블 추출\n",
    "counts = counts[metadata[metadata['Condition']!='NA'].index]\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19435688-5441-4559-9627-145cea116bbd",
   "metadata": {},
   "source": [
    "## Data filtering\n",
    "* 발현량이 없는 유전자를 필터링합니다.\n",
    "* 코멘트: %로 필터링해보기 (논문참고)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6186deab-f7a0-48cd-827b-be4ef22ef146",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "필터링 전 테이블: (36694, 76)\n",
      "필터링 후 테이블: (22262, 76)\n"
     ]
    }
   ],
   "source": [
    "# 전체 샘플에서 발현량이 없는 유전자를 필터링합니다.\n",
    "print(f'필터링 전 테이블: {counts.shape}')\n",
    "counts = counts[counts.sum(axis=1)>0]\n",
    "print(f'필터링 후 테이블: {counts.shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b244b9f6-c7bd-474a-b4c8-c633b76cd906",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "논문기반 필터링 후 테이블: (15871, 76)\n"
     ]
    }
   ],
   "source": [
    "# 논문 상 필터링 기준\n",
    "counts = counts[counts.sum(axis=1)>10]\n",
    "print(f'논문기반 필터링 후 테이블: {counts.shape}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "bbc1c397-2244-4bb2-8678-26ea82cc0b0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Symbol</th>\n",
       "      <th>TSPAN6</th>\n",
       "      <th>TNMD</th>\n",
       "      <th>DPM1</th>\n",
       "      <th>SCYL3</th>\n",
       "      <th>C1orf112</th>\n",
       "      <th>FGR</th>\n",
       "      <th>CFH</th>\n",
       "      <th>FUCA2</th>\n",
       "      <th>GCLC</th>\n",
       "      <th>NFYA</th>\n",
       "      <th>...</th>\n",
       "      <th>SND1.IT1</th>\n",
       "      <th>TSTD3</th>\n",
       "      <th>PWAR5</th>\n",
       "      <th>NDUFAF2P1</th>\n",
       "      <th>WASH9P</th>\n",
       "      <th>SOWAHCP2</th>\n",
       "      <th>KMT5AP1</th>\n",
       "      <th>BCLAF1P2</th>\n",
       "      <th>RPL23AP61</th>\n",
       "      <th>UCKL1.AS1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>X_10</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>34</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_11</th>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>30</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_13</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_14</th>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>18</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>18</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>25</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X_15</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>17</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>7</td>\n",
       "      <td>15</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 15871 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Symbol  TSPAN6  TNMD  DPM1  SCYL3  C1orf112  FGR  CFH  FUCA2  GCLC  NFYA  ...  \\\n",
       "X_10         6     0    20     18         1    3   34      8    11     3  ...   \n",
       "X_11         8     1    35     10         2    1   30      2    17     2  ...   \n",
       "X_13        10     0    14      6         1    1   35      2    15     3  ...   \n",
       "X_14         6     0    22     18         1    0   11      4    18     5  ...   \n",
       "X_15         4     1    17      9         2    0   27      7    15     4  ...   \n",
       "\n",
       "Symbol  SND1.IT1  TSTD3  PWAR5  NDUFAF2P1  WASH9P  SOWAHCP2  KMT5AP1  \\\n",
       "X_10           1      1     24          0       2         0        1   \n",
       "X_11           2      0     34          1       5         0        0   \n",
       "X_13           1      0     19          0       2         0        0   \n",
       "X_14           1      1     25          0       1         0        1   \n",
       "X_15           1      1     14          1       3         0        0   \n",
       "\n",
       "Symbol  BCLAF1P2  RPL23AP61  UCKL1.AS1  \n",
       "X_10           2          0          4  \n",
       "X_11           0          0          2  \n",
       "X_13           0          0          0  \n",
       "X_14           0          0          2  \n",
       "X_15           2          0          2  \n",
       "\n",
       "[5 rows x 15871 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 행은 샘플명, 열은 유전자 명으로 transpose\n",
    "counts = counts.T\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4f21d921-3a59-49c6-a3eb-3818c4ce2423",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/soheepark/dsc/DSC-human/RNA-LYJ/Analysis'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8822247e-79f4-488a-b519-0ceea846a6c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw 저장\n",
    "# metadata.to_csv('./GSE167186_metadata.csv', sep='\\t')\n",
    "# counts.to_csv('./GSE167186_counts.csv', sep='\\t')\n",
    "# counts.to_csv('./GSE167186_counts_filtered.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ca255c4-0ce5-433f-97b9-0dccbbb00f66",
   "metadata": {},
   "source": [
    "# DEG analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af661929-00e6-41ce-8913-20b468adf817",
   "metadata": {},
   "source": [
    "**사전 설치**:  \n",
    "`pip install pydeseq2`  \n",
    "`pip install sanbomics`  \n",
    "\n",
    "**Requirement**:  \n",
    "아래 사이트에서 pydeseq2의 설치를 위한 패키지 의존성을 확인합니다.  \n",
    "https://pydeseq2.readthedocs.io/en/latest/usage/requirements.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d382ada2-4ad5-479d-a55f-9088552e1764",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pydeseq2\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats\n",
    "from pydeseq2.utils import load_example_data\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acac3bf5-66e5-447c-adfb-baa2a1a34d78",
   "metadata": {},
   "source": [
    "## DeseqDataSet  \n",
    "\n",
    "https://pydeseq2.readthedocs.io/en/latest/api/docstrings/pydeseq2.dds.DeseqDataSet.html \n",
    "\n",
    "* DeseqDataSet은 AnnData 클래스를 확장하여 사용\n",
    "* 분산 및 logFC은 DESeq2 파이프라인에 따라 추정됨"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a716b38d-7d3a-44e5-85c0-965db11b1114",
   "metadata": {},
   "source": [
    "### DEG analysis (pipeline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3d13d07e-07b3-49ff-ad79-3b68a12276e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# raw 불러오기\n",
    "counts = pd.read_csv('./GSE167186_counts_filtered.csv', sep='\\t', index_col=[0], low_memory=False)\n",
    "metadata = pd.read_csv('./GSE167186_metadata.csv', sep='\\t', index_col=[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a1b66edf-0026-4480-bd54-2c81c435c9c8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def my_deseq(metadata, counts, name):\n",
    "    # 1. Build DeseqDataSet\n",
    "    dds = DeseqDataSet(counts=counts,\n",
    "                       metadata=metadata,\n",
    "                       design_factors=\"Condition\")\n",
    "    \n",
    "    # 2. Run DEseq2\n",
    "    dds.deseq2()\n",
    "    \n",
    "    # 3. Deseq2Stat\n",
    "    group1 = dds.obs.Condition.unique()[0]\n",
    "    group2 = dds.obs.Condition.unique()[1]\n",
    "    print(f\"comparison between {group1} and {group2}\")\n",
    "    \n",
    "    stat_res = DeseqStats(dds, contrast=('Condition', group1, group2))\n",
    "    stat_res.summary()\n",
    "    res = stat_res.results_df\n",
    "    \n",
    "    # 4. Save dds and res as pickle and csv, respectively\n",
    "    with open(f\"./{name.split('_')[1]}/results_{name}_dds.pkl\", 'wb') as f:\n",
    "        pickle.dump(dds, f)\n",
    "    res.to_csv(f\"./{name.split('_')[1]}/results_{name}.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1006881e-a97a-40e2-8d39-3fef8b20e5b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 4.12 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.53 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 5.89 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 2.85 seconds.\n",
      "\n",
      "Replacing 32 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['OH' 'YH']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.29 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Condition OH vs YH\n",
      "            baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "TSPAN6      5.465733       -0.086127  0.199698 -0.431287  0.666260  0.895437\n",
      "TNMD        0.523736        0.201461  0.726083  0.277463  0.781424       NaN\n",
      "DPM1       22.466529       -0.005716  0.131542 -0.043457  0.965337  0.991708\n",
      "SCYL3      10.713833       -0.188360  0.147720 -1.275108  0.202271  0.564170\n",
      "C1orf112    0.675763        0.305825  0.609216  0.501998  0.615669       NaN\n",
      "...              ...             ...       ...       ...       ...       ...\n",
      "SOWAHCP2    0.255236        0.028924  0.925358  0.031257  0.975064       NaN\n",
      "KMT5AP1     0.174994        0.385016  1.222162  0.315029  0.752740       NaN\n",
      "BCLAF1P2    0.511030       -0.099830  0.644070 -0.154998  0.876823       NaN\n",
      "RPL23AP61   0.356363       -0.569868  0.782595 -0.728177  0.466505       NaN\n",
      "UCKL1.AS1   2.053895        0.180704  0.361389  0.500026  0.617057  0.874972\n",
      "\n",
      "[15871 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 6.24 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.76 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 9.69 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 3.70 seconds.\n",
      "\n",
      "Replacing 20 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Running Wald tests...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['OS' 'YH']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.46 seconds.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Condition OS vs YH\n",
      "            baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "TSPAN6      5.803493        0.070755  0.209828  0.337206  0.735961  0.891657\n",
      "TNMD        0.437790       -0.170323  0.867450 -0.196349  0.844337       NaN\n",
      "DPM1       22.086787       -0.079989  0.138211 -0.578750  0.562758  0.799846\n",
      "SCYL3      10.471977       -0.292125  0.156967 -1.861055  0.062736  0.250261\n",
      "C1orf112    0.529970       -0.229913  0.668097 -0.344131  0.730747       NaN\n",
      "...              ...             ...       ...       ...       ...       ...\n",
      "SOWAHCP2    0.215972       -0.177358  1.078048 -0.164518  0.869324       NaN\n",
      "KMT5AP1     0.182329        0.518679  1.296751  0.399983  0.689169       NaN\n",
      "BCLAF1P2    0.412857       -0.695597  0.749499 -0.928083  0.353364       NaN\n",
      "RPL23AP61   0.387560       -0.417041  0.783798 -0.532077  0.594672       NaN\n",
      "UCKL1.AS1   1.732130       -0.138994  0.418908 -0.331801  0.740039  0.892543\n",
      "\n",
      "[15871 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 5.19 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.54 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 6.79 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 3.26 seconds.\n",
      "\n",
      "Replacing 26 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Running Wald tests...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['OS' 'OH']\n",
      "Log2 fold change & Wald test p-value: Condition OS vs OH\n",
      "            baseMean  log2FoldChange     lfcSE      stat    pvalue  padj\n",
      "TSPAN6      5.654878        0.157117  0.188042  0.835542  0.403413   NaN\n",
      "TNMD        0.473517       -0.374364  0.669769 -0.558946  0.576199   NaN\n",
      "DPM1       22.147004       -0.073913  0.122217 -0.604773  0.545330   NaN\n",
      "SCYL3       9.969259       -0.103497  0.142652 -0.725516  0.468136   NaN\n",
      "C1orf112    0.626903       -0.535939  0.597332 -0.897222  0.369600   NaN\n",
      "...              ...             ...       ...       ...       ...   ...\n",
      "SOWAHCP2    0.237148       -0.205204  0.959586 -0.213846  0.830667   NaN\n",
      "KMT5AP1     0.238784        0.134741  0.962118  0.140047  0.888623   NaN\n",
      "BCLAF1P2    0.404395       -0.591307  0.696319 -0.849190  0.395776   NaN\n",
      "RPL23AP61   0.304366        0.154205  0.815042  0.189198  0.849937   NaN\n",
      "UCKL1.AS1   1.937287       -0.320551  0.366878 -0.873727  0.382267   NaN\n",
      "\n",
      "[15871 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.17 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# 3개 비교군 나누기\n",
    "metadata_ohyh = metadata[(metadata['Condition']=='OH') | (metadata['Condition']=='YH')]\n",
    "metadata_osoh = metadata[(metadata['Condition']=='OS') | (metadata['Condition']=='OH')]\n",
    "metadata_osyh = metadata[(metadata['Condition']=='OS') | (metadata['Condition']=='YH')]\n",
    "\n",
    "counts_ohyh = counts.loc[metadata_ohyh.index]\n",
    "counts_osoh = counts.loc[metadata_osoh.index]\n",
    "counts_osyh = counts.loc[metadata_osyh.index]\n",
    "\n",
    "# my_deseq 구동\n",
    "my_deseq(metadata_ohyh, counts_ohyh, 'GSE167186_ohyh_filtered')\n",
    "my_deseq(metadata_osoh, counts_osoh, 'GSE167186_osoh_filtered')\n",
    "my_deseq(metadata_osyh, counts_osyh, 'GSE167186_osyh_filtered')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "084a3bfc-26cf-49a4-9809-b11f9565c84c",
   "metadata": {},
   "source": [
    "# GSEA\n",
    "* (코멘트) gsea html 파일이 나옴! gmt, tms, gct 파일 input\n",
    "* GO에 대한 유전자 score에 따른 정보 나옴"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f536e7a-3511-40f4-841f-cc00b0f1fe7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "res.head()\n",
    "res['Symbol'] = res.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "389cc62d-8b23-4a63-aed0-4b3405cbe98c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ranking = res[['Symbol','stat']].dropna().sort_values('stat', ascending=False)\n",
    "ranking = ranking.drop_duplicates('Symbol')\n",
    "ranking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8f71908-a37c-4552-8f0d-5cacc7141aac",
   "metadata": {},
   "outputs": [],
   "source": [
    "manual_set = {'things':['COL19A1', 'H19', 'FGF7', 'NHLH2', 'MT2A']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e05e2f89-281c-4a0f-b551-f3d9bb52cfea",
   "metadata": {},
   "outputs": [],
   "source": [
    "del ranking['Symbol']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "712015cf-3ece-49ae-b2d3-115f6d66f96c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_res = gp.prerank(rnk = ranking,\n",
    "                     gene_sets = ['GO_Biological_Process_2021', manual_set],\n",
    "                     seed = 6, permutation_num = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9386180a-a771-447c-beaa-750c70a6a5a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "out = []\n",
    "\n",
    "for term in list(pre_res.results):\n",
    "    out.append([term,\n",
    "               pre_res.results[term]['fdr'],\n",
    "               pre_res.results[term]['es'],\n",
    "               pre_res.results[term]['nes']])\n",
    "\n",
    "out_df = pd.DataFrame(out, columns = ['Term','fdr', 'es', 'nes']).sort_values('fdr').reset_index(drop = True)\n",
    "out_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cacda5f-60bc-4509-a7b3-4bd71df5d2e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(out_df.sort_values('nes').iloc[0].Term)\n",
    "print(out_df.sort_values('nes').iloc[1].Term)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c717894d-ad6f-4c24-aed5-b2af4aa0712d",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_term = out_df.sort_values('nes').iloc[0].Term\n",
    "my_term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f47038c-388b-4a33-8d95-beffe0a213e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gseaplot(rank_metric=pre_res.ranking, term=my_term, **pre_res.results[my_term])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b6335f4-13a9-4c48-ae30-28fdad0bb046",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Build DeseqDataSet "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "c96bd134-51a6-4e96-ac96-acf3ff0d36b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['OS', 'OH', 'Un', 'YH'], dtype=object)"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['Condition_'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "543316ae-68d9-42a3-b75b-47faeb3a17e7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/zm/q_hkpfmd4c52bmx280q27v980000gn/T/ipykernel_6932/3494598582.py:1: UserWarning: Same factor names in the design contain underscores ('_'). They will\n",
      "                be converted to hyphens ('-').\n",
      "  dds = DeseqDataSet(counts=counts.loc[metadata[metadata['Condition_']!='Un'].index],\n",
      "Fitting size factors...\n",
      "... done in 0.05 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 4.94 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.71 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 7.95 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 4.01 seconds.\n",
      "\n",
      "Replacing 29 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds = DeseqDataSet(counts=counts.loc[metadata[metadata['Condition_']!='Un'].index],\n",
    "                   metadata=metadata[metadata['Condition_']!='Un'],\n",
    "                   design_factors=\"Condition_\")\n",
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "09053fa5-6cd3-4fae-adb2-b540d91918dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 72 × 22262\n",
       "    obs: 'Condition', 'Condition-', 'Sample', 'Temp'\n",
       "    uns: 'trend_coeffs', 'disp_function_type', '_squared_logres', 'prior_disp_var'\n",
       "    obsm: 'design_matrix', 'size_factors', 'replaceable'\n",
       "    varm: 'non_zero', '_MoM_dispersions', 'genewise_dispersions', '_genewise_converged', '_normed_means', 'fitted_dispersions', 'MAP_dispersions', '_MAP_converged', 'dispersions', '_outlier_genes', 'LFC', '_LFC_converged', 'replaced', 'refitted'\n",
       "    layers: 'normed_counts', '_mu_hat', '_mu_LFC', '_hat_diagonals', 'cooks', 'replace_cooks'"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "bfca02b2-c984-44cc-8144-d3d93d2e01cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Old Sarcopenia', 'Old Healthy', 'Young Healthy'], dtype=object)"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.obs.Condition.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8107b45-4d54-4d4f-bf90-f749998c8795",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Run DESeq2  \n",
    "분산과 logFC를 fit하기 위해 DESeq2 파이프라인에 따라 추정됩니다. 아래와 같은 항목이 계산되어 anndata에 추가됩니다.\n",
    "* Compute normalization factors\n",
    "* Fit dispersion trend coefficients\n",
    "* Dispersion priors\n",
    "* MAP Dispersions\n",
    "* Fit log fold changes\n",
    "* Calculate Cooks distances and refit (Optional)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "51a0bfdb-9d7f-486d-8531-dd65342c4e6d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 4.42 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.47 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 6.06 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 3.14 seconds.\n",
      "\n",
      "Replacing 20 outlier genes.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# deseq 계산\n",
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "607ca876-849c-41e9-bda3-287412fbe8bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 72 × 22262\n",
       "    obs: 'Condition', 'Condition-', 'Sample', 'Temp'\n",
       "    uns: 'trend_coeffs', 'disp_function_type', '_squared_logres', 'prior_disp_var'\n",
       "    obsm: 'design_matrix', 'size_factors', 'replaceable'\n",
       "    varm: 'non_zero', '_MoM_dispersions', 'genewise_dispersions', '_genewise_converged', '_normed_means', 'fitted_dispersions', 'MAP_dispersions', '_MAP_converged', 'dispersions', '_outlier_genes', 'LFC', '_LFC_converged', 'replaced', 'refitted'\n",
       "    layers: 'normed_counts', '_mu_hat', '_mu_LFC', '_hat_diagonals', 'cooks', 'replace_cooks'"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed88f871-9a6d-49bf-a4a5-f9bc45a355ac",
   "metadata": {},
   "source": [
    "Anndata에 다음 항목이 추가됩니다.  \n",
    "* `X`: the count data\n",
    "* `obs`: design factors\n",
    "* `obsm`: sample-level data (ex: design_matrix, size_factors)\n",
    "* `varm`: gene-level data (ex: dispersions, LFC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6ddb4b20-cd89-48f6-b8e9-4fdcc8db676e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Raw count 및 Normalized count 유전자 발현량 분포 확인\n",
    "\n",
    "# # 각 열의 합 또는 평균을 계산\n",
    "# column_raw = dds.X.mean(axis=1)  # Raw count 평균\n",
    "# column_norm = dds.layers['normed_counts'].mean(axis=1)  # Normazlied 평균\n",
    "\n",
    "# # 합 또는 평균의 분포를 히스토그램으로 시각화\n",
    "# plt.figure(figsize=(10, 5))\n",
    "\n",
    "# # Raw count 평균 히스토그램\n",
    "# plt.subplot(1, 2, 1)\n",
    "# sns.histplot(column_raw, kde=True, color=\"blue\", bins=30)\n",
    "# plt.title(\"Mean of Raw Counts Distribution\")\n",
    "\n",
    "# # Normazlied 평균 히스토그램\n",
    "# plt.subplot(1, 2, 2)\n",
    "# sns.histplot(column_norm, kde=True, color=\"green\", bins=30)\n",
    "# plt.title(\"Mean of Normalized Counts Distribution\")\n",
    "\n",
    "# plt.tight_layout()\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54c9afed-1231-4dde-a952-65560c31e3aa",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## DeseqStats\n",
    "\n",
    "https://pydeseq2.readthedocs.io/en/latest/api/docstrings/pydeseq2.ds.DeseqStats.html\n",
    "\n",
    "* 미분 표현을 위한 PyDESeq2 통계 테스트\n",
    "* DESeq2 파이프라인에 따라 차등 유전자 발현에 대한 p-value 추정\n",
    "* apeGLM logFC 축소 지원"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "eabb3698-b066-46ba-976b-d65964c45a79",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: Condition Sarcopenia vs Young Healthy\n",
      "            baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
      "Symbol                                                                      \n",
      "TSPAN6      5.803493        0.070761  0.209034  0.338517  0.734974  0.891126\n",
      "TNMD        0.437790       -0.170577  0.861031 -0.198108  0.842961       NaN\n",
      "DPM1       22.086787       -0.080009  0.138125 -0.579249  0.562421  0.799059\n",
      "SCYL3      10.471977       -0.292301  0.156472 -1.868065  0.061753  0.246922\n",
      "C1orf112    0.529970       -0.229904  0.664676 -0.345890  0.729426       NaN\n",
      "...              ...             ...       ...       ...       ...       ...\n",
      "GATD3B      0.027959       -0.171026  3.075026 -0.055618  0.955646       NaN\n",
      "PCDH20      0.022076       -0.171026  3.075026 -0.055618  0.955646       NaN\n",
      "UCKL1.AS1   1.732130       -0.138432  0.416866 -0.332078  0.739830  0.891747\n",
      "FRG1KP      0.000000             NaN       NaN       NaN       NaN       NaN\n",
      "ERICD       0.117460        0.581956  3.058409  0.190281  0.849089       NaN\n",
      "\n",
      "[22126 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... done in 1.35 seconds.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "stat_res = DeseqStats(dds, contrast=('Condition', \n",
    "                                     dds.obs.Condition.unique()[0], \n",
    "                                     dds.obs.Condition.unique()[1]))\n",
    "stat_res.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "68953347-018c-4207-9278-2f6070c8dc39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>5.803493</td>\n",
       "      <td>0.070761</td>\n",
       "      <td>0.209034</td>\n",
       "      <td>0.338517</td>\n",
       "      <td>0.734974</td>\n",
       "      <td>0.891126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>0.437790</td>\n",
       "      <td>-0.170577</td>\n",
       "      <td>0.861031</td>\n",
       "      <td>-0.198108</td>\n",
       "      <td>0.842961</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>22.086787</td>\n",
       "      <td>-0.080009</td>\n",
       "      <td>0.138125</td>\n",
       "      <td>-0.579249</td>\n",
       "      <td>0.562421</td>\n",
       "      <td>0.799059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>10.471977</td>\n",
       "      <td>-0.292301</td>\n",
       "      <td>0.156472</td>\n",
       "      <td>-1.868065</td>\n",
       "      <td>0.061753</td>\n",
       "      <td>0.246922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>0.529970</td>\n",
       "      <td>-0.229904</td>\n",
       "      <td>0.664676</td>\n",
       "      <td>-0.345890</td>\n",
       "      <td>0.729426</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATD3B</th>\n",
       "      <td>0.027959</td>\n",
       "      <td>-0.171026</td>\n",
       "      <td>3.075026</td>\n",
       "      <td>-0.055618</td>\n",
       "      <td>0.955646</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCDH20</th>\n",
       "      <td>0.022076</td>\n",
       "      <td>-0.171026</td>\n",
       "      <td>3.075026</td>\n",
       "      <td>-0.055618</td>\n",
       "      <td>0.955646</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UCKL1.AS1</th>\n",
       "      <td>1.732130</td>\n",
       "      <td>-0.138432</td>\n",
       "      <td>0.416866</td>\n",
       "      <td>-0.332078</td>\n",
       "      <td>0.739830</td>\n",
       "      <td>0.891747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FRG1KP</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERICD</th>\n",
       "      <td>0.117460</td>\n",
       "      <td>0.581956</td>\n",
       "      <td>3.058409</td>\n",
       "      <td>0.190281</td>\n",
       "      <td>0.849089</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22126 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "Symbol                                                                      \n",
       "TSPAN6      5.803493        0.070761  0.209034  0.338517  0.734974  0.891126\n",
       "TNMD        0.437790       -0.170577  0.861031 -0.198108  0.842961       NaN\n",
       "DPM1       22.086787       -0.080009  0.138125 -0.579249  0.562421  0.799059\n",
       "SCYL3      10.471977       -0.292301  0.156472 -1.868065  0.061753  0.246922\n",
       "C1orf112    0.529970       -0.229904  0.664676 -0.345890  0.729426       NaN\n",
       "...              ...             ...       ...       ...       ...       ...\n",
       "GATD3B      0.027959       -0.171026  3.075026 -0.055618  0.955646       NaN\n",
       "PCDH20      0.022076       -0.171026  3.075026 -0.055618  0.955646       NaN\n",
       "UCKL1.AS1   1.732130       -0.138432  0.416866 -0.332078  0.739830  0.891747\n",
       "FRG1KP      0.000000             NaN       NaN       NaN       NaN       NaN\n",
       "ERICD       0.117460        0.581956  3.058409  0.190281  0.849089       NaN\n",
       "\n",
       "[22126 rows x 6 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res = stat_res.results_df\n",
    "res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "742acc99-4f24-45c2-bec9-6f93186b95b4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
